File: fabric/1.21.8/src/client/java/dev/errnicraft/chatremastered/ChatRemasteredContextMenu.java
package dev.errnicraft.chatremastered;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import org.lwjgl.glfw.GLFW;

import java.util.List;

public class ChatRemasteredContextMenu extends Screen {

    public static final class MenuItem {
        private final String label;
        private final String icon;
        private final Runnable action;

        public MenuItem(String label, String icon, Runnable action) {
            this.label = label;
            this.icon = icon;
            this.action = action;
        }

        public MenuItem(String label, Runnable action) {
            this(label, "", action);
        }

        public String getLabel() {
            return label;
        }

        public String getIcon() {
            return icon;
        }

        public Runnable getAction() {
            return action;
        }
    }

    private final List<MenuItem> items;
    private final int anchorX;
    private final int anchorY;
    private final Screen parent;

    public ChatRemasteredContextMenu(List<MenuItem> items, int anchorX, int anchorY, Screen parent) {
        super(Component.empty());
        this.items = items;
        this.anchorX = anchorX;
        this.anchorY = anchorY;
        this.parent = parent;
        this.itemColors = new int[items.size()];
    }

    public static ChatRemasteredContextMenu forImage(String imageId, int ax, int ay, Screen parent) {
        return new ChatRemasteredContextMenu(
                List.of(
                        new MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_save_as"), "💾", () ->
                                ChatRemasteredClient.saveImageAs(imageId)),
                        new MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_copy_id"), "📋", () -> {
                            Minecraft mc = Minecraft.getInstance();
                            mc.keyboardHandler.setClipboard(imageId);
                            mc.gui.getChat().addMessage(
                                    Component.literal("§8[Chat Remastered] §7" + ChatRemasteredConfig.tr("chat-remastered.id_copied", imageId, imageId))
                            );
                        })
                ), ax, ay, parent
        );
    }

    public static ChatRemasteredContextMenu forMessage(String text, int ax, int ay, Screen parent) {
        return new ChatRemasteredContextMenu(
                List.of(
                        new MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_copy_message"), "📋", () ->
                                Minecraft.getInstance().keyboardHandler.setClipboard(text))
                ), ax, ay, parent
        );
    }

    private final int ITEM_W = 180;
    private final int ITEM_H = 20;
    private final int PAD = 6;
    private final float RADIUS = 6f;
    private final long ANIM_MS = 120L;

    private int menuX = 0;
    private int menuY = 0;
    private long openTime = 0L;
    private int hoveredIndex = -1;

    private final int[] itemColors;

    @Override
    protected void init() {
        openTime = System.currentTimeMillis();
        int menuH = PAD * 2 + items.size() * ITEM_H + (items.size() - 1) * 2;
        menuX = Math.min(anchorX, width - ITEM_W - PAD * 2);
        menuY = Math.min(anchorY, height - menuH - PAD * 2);
    }

    @Override
    public void render(GuiGraphics graphics, int mouseX, int mouseY, float delta) {
        Minecraft mc = Minecraft.getInstance();
        int menuH = PAD * 2 + items.size() * ITEM_H + (items.size() - 1) * 2;

        float elapsed = (System.currentTimeMillis() - openTime);
        float t = Math.min(elapsed / ANIM_MS, 1f);
        float ease = 1f - (1f - t) * (1f - t);

        graphics.pose().pushMatrix();
        float cx = menuX + ITEM_W / 2f;
        float cy = menuY;
        graphics.pose().translate(cx, cy);
        graphics.pose().scale(ease, ease);
        graphics.pose().translate(-cx, -cy);

        graphics.fill(menuX + 3, menuY + 3, menuX + ITEM_W + 3, menuY + menuH + 3, 0x44000000);
        graphics.fill(menuX + 2, menuY + 2, menuX + ITEM_W + 2, menuY + menuH + 2, 0x22000000);

        graphics.fill(menuX, menuY, menuX + ITEM_W, menuY + menuH, (int) 0xF52C2C2C);
        graphics.fill(menuX, menuY, menuX + ITEM_W, menuY + 1, (int) 0xFF555555);
        graphics.fill(menuX, menuY + menuH - 1, menuX + ITEM_W, menuY + menuH, (int) 0xFF555555);
        graphics.fill(menuX, menuY, menuX + 1, menuY + menuH, (int) 0xFF555555);
        graphics.fill(menuX + ITEM_W - 1, menuY, menuX + ITEM_W, menuY + menuH, (int) 0xFF555555);

        hoveredIndex = -1;
        for (int i = 0; i < items.size(); i++) {
            MenuItem item = items.get(i);
            int iy = menuY + PAD + i * (ITEM_H + 2);
            boolean hovered = mouseX >= menuX + 1 && mouseX < menuX + ITEM_W - 1 && mouseY >= iy && mouseY < iy + ITEM_H;

            if (hovered) {
                hoveredIndex = i;
            }

            int targetBg = hovered ? (int) 0xFF3D6099 : 0x002C2C2C;
            itemColors[i] = lerpColor(itemColors[i], targetBg, 0.25f);
            if (hovered || (itemColors[i] >>> 24) > 5) {
                graphics.fill(menuX + 2, iy, menuX + ITEM_W - 2, iy + ITEM_H, itemColors[i]);
            }

            int textColor = hovered ? (int) 0xFFFFFFFF : (int) 0xFFDDDDDD;
            int textX = menuX + PAD + (!item.getIcon().isEmpty() ? 14 : 0);
            int textY = iy + (ITEM_H - mc.font.lineHeight) / 2;
            if (!item.getIcon().isEmpty()) {
                graphics.drawString(mc.font, item.getIcon(), menuX + PAD - 1, textY, textColor, false);
            }
            graphics.drawString(mc.font, item.getLabel(), textX, textY, textColor, false);
        }

        graphics.pose().popMatrix();
    }

    private int lerpColor(int from, int to, float t) {
        int fa = (from >>> 24) & 0xFF;
        int fr = (from >>> 16) & 0xFF;
        int fg = (from >>> 8) & 0xFF;
        int fb = from & 0xFF;
        int ta = (to >>> 24) & 0xFF;
        int tr = (to >>> 16) & 0xFF;
        int tg = (to >>> 8) & 0xFF;
        int tb = to & 0xFF;
        int a = clamp((int) (fa + (ta - fa) * t), 0, 255);
        int r = clamp((int) (fr + (tr - fr) * t), 0, 255);
        int g = clamp((int) (fg + (tg - fg) * t), 0, 255);
        int b = clamp((int) (fb + (tb - fb) * t), 0, 255);
        return (a << 24) | (r << 16) | (g << 8) | b;
    }

    private int clamp(int v, int min, int max) {
        return Math.max(min, Math.min(max, v));
    }

    @Override
    public boolean mouseClicked(double mouseXD, double mouseYD, int button) {
        int mouseX = (int) mouseXD;
        int mouseY = (int) mouseYD;
        int menuH = PAD * 2 + items.size() * ITEM_H + (items.size() - 1) * 2;
        boolean inside = mouseX >= menuX && mouseX <= menuX + ITEM_W && mouseY >= menuY && mouseY <= menuY + menuH;
        if (!inside) {
            minecraft.setScreen(parent);
            return true;
        }
        for (int i = 0; i < items.size(); i++) {
            MenuItem item = items.get(i);
            int iy = menuY + PAD + i * (ITEM_H + 2);
            if (mouseX >= menuX + 1 && mouseX < menuX + ITEM_W - 1 && mouseY >= iy && mouseY < iy + ITEM_H) {
                item.getAction().run();
                minecraft.setScreen(parent);
                return true;
            }
        }
        return true;
    }

    @Override
    public boolean keyPressed(int keyCode, int scanCode, int modifiers) {
        if (keyCode == GLFW.GLFW_KEY_ESCAPE) {
            minecraft.setScreen(parent);
            return true;
        }
        return super.keyPressed(keyCode, scanCode, modifiers);
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }

    @Override
    public boolean shouldCloseOnEsc() {
        return false;
    }
}

File: src/client/kotlin/dev/errnicraft/chatremastered/ChatRemasteredContextMenu.kt
package dev.errnicraft.chatremastered

import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.screens.Screen
import net.minecraft.network.chat.Component
import org.lwjgl.glfw.GLFW

class ChatRemasteredContextMenu(
    private val items: List<MenuItem>,
    private val anchorX: Int,
    private val anchorY: Int,
    private val parent: Screen
) : Screen(Component.empty()) {

    data class MenuItem(val label: String, val icon: String = "", val action: () -> Unit)

    companion object {
        @JvmStatic
        fun forImage(imageId: String, ax: Int, ay: Int, parent: Screen) = ChatRemasteredContextMenu(
            listOf(
                MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_save_as"), "💾") {
                    ChatRemasteredClient.saveImageAs(imageId)
                },
                MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_copy_id"), "📋") {
                    val mc = net.minecraft.client.Minecraft.getInstance()
                    mc.keyboardHandler.setClipboard(imageId)
                    mc.gui.chat.addMessage(
                        Component.literal("§8[Chat Remastered] §7" + ChatRemasteredConfig.tr("chat-remastered.id_copied", imageId, imageId))
                    )
                }
            ), ax, ay, parent
        )

        @JvmStatic
        fun forMessage(text: String, ax: Int, ay: Int, parent: Screen) = ChatRemasteredContextMenu(
            listOf(
                MenuItem(ChatRemasteredConfig.tr("chat-remastered.ctx_copy_message"), "📋") {
                    net.minecraft.client.Minecraft.getInstance().keyboardHandler.setClipboard(text)
                }
            ), ax, ay, parent
        )
    }

    private val ITEM_W = 180
    private val ITEM_H = 20
    private val PAD = 6
    private val RADIUS = 6f
    private val ANIM_MS = 120L

    private var menuX = 0
    private var menuY = 0
    private var openTime = 0L
    private var hoveredIndex = -1

    private val itemColors = IntArray(items.size) { 0 }

    override fun init() {
        openTime = System.currentTimeMillis()
        val menuH = PAD * 2 + items.size * ITEM_H + (items.size - 1) * 2
        menuX = anchorX.coerceAtMost(width - ITEM_W - PAD * 2)
        menuY = anchorY.coerceAtMost(height - menuH - PAD * 2)
    }

    override fun render(graphics: GuiGraphics, mouseX: Int, mouseY: Int, delta: Float) {
        val mc = net.minecraft.client.Minecraft.getInstance()
        val menuH = PAD * 2 + items.size * ITEM_H + (items.size - 1) * 2

        val elapsed = (System.currentTimeMillis() - openTime).toFloat()
        val t = (elapsed / ANIM_MS).coerceAtMost(1f)
        val ease = 1f - (1f - t) * (1f - t)

        graphics.pose().pushMatrix()
        val cx = (menuX + ITEM_W / 2).toFloat()
        val cy = menuY.toFloat()
        graphics.pose().translate(cx, cy)
        graphics.pose().scale(ease, ease)
        graphics.pose().translate(-cx, -cy)

        // Тень
        graphics.fill(menuX + 3, menuY + 3, menuX + ITEM_W + 3, menuY + menuH + 3, 0x44000000)
        graphics.fill(menuX + 2, menuY + 2, menuX + ITEM_W + 2, menuY + menuH + 2, 0x22000000)

        // Фон панели
        graphics.fill(menuX, menuY, menuX + ITEM_W, menuY + menuH, 0xF52C2C2C.toInt())
        // Рамка
        graphics.fill(menuX, menuY, menuX + ITEM_W, menuY + 1, 0xFF555555.toInt())
        graphics.fill(menuX, menuY + menuH - 1, menuX + ITEM_W, menuY + menuH, 0xFF555555.toInt())
        graphics.fill(menuX, menuY, menuX + 1, menuY + menuH, 0xFF555555.toInt())
        graphics.fill(menuX + ITEM_W - 1, menuY, menuX + ITEM_W, menuY + menuH, 0xFF555555.toInt())

        hoveredIndex = -1
        items.forEachIndexed { i, item ->
            val iy = menuY + PAD + i * (ITEM_H + 2)
            val hovered = mouseX >= menuX + 1 && mouseX < menuX + ITEM_W - 1 && mouseY >= iy && mouseY < iy + ITEM_H

            if (hovered) hoveredIndex = i

            val targetBg = if (hovered) 0xFF3D6099.toInt() else 0x002C2C2C
            itemColors[i] = lerpColor(itemColors[i], targetBg, 0.25f)
            if (hovered || itemColors[i].ushr(24) > 5) {
                graphics.fill(menuX + 2, iy, menuX + ITEM_W - 2, iy + ITEM_H, itemColors[i])
            }

            val textColor = if (hovered) 0xFFFFFFFF.toInt() else 0xFFDDDDDD.toInt()
            val textX = menuX + PAD + if (item.icon.isNotEmpty()) 14 else 0
            val textY = iy + (ITEM_H - mc.font.lineHeight) / 2
            if (item.icon.isNotEmpty()) {
                graphics.drawString(mc.font, item.icon, menuX + PAD - 1, textY, textColor, false)
            }
            graphics.drawString(mc.font, item.label, textX, textY, textColor, false)
        }

        graphics.pose().popMatrix()
    }

    private fun lerpColor(from: Int, to: Int, t: Float): Int {
        val fa = (from ushr 24) and 0xFF
        val fr = (from ushr 16) and 0xFF
        val fg = (from ushr 8) and 0xFF
        val fb = from and 0xFF
        val ta = (to ushr 24) and 0xFF
        val tr = (to ushr 16) and 0xFF
        val tg = (to ushr 8) and 0xFF
        val tb = to and 0xFF
        val a = (fa + (ta - fa) * t).toInt().coerceIn(0, 255)
        val r = (fr + (tr - fr) * t).toInt().coerceIn(0, 255)
        val g = (fg + (tg - fg) * t).toInt().coerceIn(0, 255)
        val b = (fb + (tb - fb) * t).toInt().coerceIn(0, 255)
        return (a shl 24) or (r shl 16) or (g shl 8) or b
    }

    override fun mouseClicked(event: net.minecraft.client.input.MouseButtonEvent, doubleClick: Boolean): Boolean {
        val mouseX = event.x().toInt()
        val mouseY = event.y().toInt()
        val menuH = PAD * 2 + items.size * ITEM_H + (items.size - 1) * 2
        val inside = mouseX >= menuX && mouseX <= menuX + ITEM_W && mouseY >= menuY && mouseY <= menuY + menuH
        if (!inside) {
            minecraft!!.setScreen(parent)
            return true
        }
        items.forEachIndexed { i, item ->
            val iy = menuY + PAD + i * (ITEM_H + 2)
            if (mouseX >= menuX + 1 && mouseX < menuX + ITEM_W - 1 && mouseY >= iy && mouseY < iy + ITEM_H) {
                item.action()
                minecraft!!.setScreen(parent)
                return true
            }
        }
        return true
    }

    override fun keyPressed(event: net.minecraft.client.input.KeyEvent): Boolean {
        if (event.key() == GLFW.GLFW_KEY_ESCAPE) {
            minecraft!!.setScreen(parent)
            return true
        }
        return super.keyPressed(event)
    }

    override fun isPauseScreen() = false
    override fun shouldCloseOnEsc() = false
}
